## Book Recomendation System 

<p> A recommendation engine is a class of machine learning which offers relevant suggestions to the customer.  Before the recommendation system, the major tendency to buy was to take a suggestion from friends. But Now Google knows what news you will read, Youtube knows what type of videos you will watch based on your search history, watch history, or purchase history.<br>

A recommendation system helps an organization to create loyal customers and build trust by them desired products and services for which they came on your site. The recommendation system today are so powerful that they can handle the new customer too who has visited the site for the first time. They recommend the products which are currently trending or highly rated and they can also recommend the products which bring maximum profit to the company.</p>

<p><i> A book recommendation system is a type of recommendation system where we have to recommend similar books to the reader based on his interest. The books recommendation system is used by online websites which provide ebooks like google play books, open library, good Read’s, etc.</i>
<p>

<br><p>In this project we will be using collabrative way filetring method to build book recomandation system</p><br><br><h2>DATASET : </h2>

we have 3 files in our dataset which is extracted from some books selling websites.
<ul>
 <li>Books – first are about books which contain all the information related to books like an author, title, publication year, etc.</li>
    <li>Users – The second file contains registered user’s information like user id, location.</li>
<li>ratings –  Ratings contain information like which user has given how much rating to which book.</li>
</ul>
So based on all these three files we can build a powerful collaborative filtering model. let’s get started. <br><br>

<h3>LOADING DATASET</h3>

<p>let us start while importing libraries and load datasets. while loading the file we have some problems like.
<ul>
 <li>The values in the CSV file are separated by semicolons, not by a comma.</li>
<li>There are some lines which not work like we cannot import it with pandas and It throws an error because python is Interpreted language.</li>
<li>Encoding of a file is in Latin<li></ul>
So while loading data we have to handle these exceptions and after running the below code you will get some warning and it will show which lines have an error that we have skipped while loading.<p>

In [1]:
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

books = pd.read_csv("BX-Books.csv", sep=';', encoding="latin-1", error_bad_lines=False)
users = pd.read_csv("BX-Users.csv", sep=';', encoding="latin-1", error_bad_lines=False)
ratings = pd.read_csv("BX-Book-Ratings.csv", sep=';', encoding="latin-1", error_bad_lines=False)

books.head(3)

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'


ISBN            Book-Title           Book-Author Year-Of-Publication  \
0  0195153448   Classical Mythology    Mark P. O. Morford                2002   
1  0002005018          Clara Callan  Richard Bruce Wright                2001   
2  0060973129  Decision in Normandy          Carlo D'Este                1991   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2          HarperPerennial  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...

In [2]:
books.keys()

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [3]:
books.keys()

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [4]:
print(users.keys())


Index(['User-ID', 'Location', 'Age'], dtype='object')


In [5]:
users.head(10)

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN
5        6       santa monica, california, usa  61.0
6        7                 washington, dc, usa   NaN
7        8            timmins, ontario, canada   NaN
8        9          germantown, tennessee, usa   NaN
9       10          albacete, wisconsin, spain  26.0

In [6]:
print(ratings.keys())

Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')


In [7]:
ratings.head(5)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [8]:
books.head(5)

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

## PreProcessing DATA

<p>Now in the books file, we have some extra columns which are not required for our task like image URLs. And we will rename the columns of each file as the name of the column contains space, and uppercase letters so we will correct as to make it easy to use.</p>

In [9]:
books = books[['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher']]
books.rename(columns = {'Book-Title':'title', 'Book-Author':'author', 'Year-Of-Publication':'year', 'Publisher':'publisher'}, inplace=True)
users.rename(columns = {'User-ID':'user_id', 'Location':'location', 'Age':'age'}, inplace=True)
ratings.rename(columns = {'User-ID':'user_id', 'Book-Rating':'rating'}, inplace=True)

In [10]:
books.head(5)

ISBN                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  
0    Mark P. O. Morford  2002     Oxford University Press  
1  Richard Bruce Wright  2001       HarperFlamingo Canada  
2          Carlo D'Este  1991             HarperPerennial  
3      Gina Bari Kolata  1999        Farrar Straus Giroux  
4       E. J. W. Barber  1999  W. W. Norton &amp; Company

In [11]:
users.head(5)

user_id                            location   age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [12]:
ratings.head(5)

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5
2   276727  0446520802       0
3   276729  052165615X       3
4   276729  0521795028       6

In [13]:
books.shape       #total book availabele

(271360, 5)

In [14]:
users.shape        # total registered user

(278858, 3)

In [15]:
ratings.shape

(1149780, 3)

<h3> Exploratory data Analysis </h3>

<p><i>we want to do that If there is user A who has read and liked x and y books, And user B has also liked this two books and now user A has read and liked some z book which is not read by B so we have to recommend z book to user B. This is what collaborative filtering is.</p><br></i>

<p>If we take all the books and all the users for modeling, it will it create a problem. So what we have to do is we have to decrease the number of users and books because we cannot consider a user who has only registered on the website or has only read one or two books. On such a user, we cannot rely to recommend books to others because we have to extract knowledge from data. So what we will limit this number and we will take a user who has rated at least 200 books and also we will limit books and we will take only those books which have received at least 50 ratings from a user.<p>

In [32]:
books.loc[0:20]

    

ISBN                                              title  \
0   0195153448                                Classical Mythology   
1   0002005018                                       Clara Callan   
2   0060973129                               Decision in Normandy   
3   0374157065  Flu: The Story of the Great Influenza Pandemic...   
4   0393045218                             The Mummies of Urumchi   
5   0399135782                             The Kitchen God's Wife   
6   0425176428  What If?: The World's Foremost Military Histor...   
7   0671870432                                    PLEADING GUILTY   
8   0679425608  Under the Black Flag: The Romance and the Real...   
9   074322678X            Where You'll Find Me: And Other Stories   
10  0771074670                        Nights Below Station Street   
11  080652121X  Hitler's Secret Bankers: The Myth of Swiss Neu...   
12  0887841740                                 The Middle Stories   
13  1552041778                                           Jane Doe   
14  1558746218  A Second Chicken Soup for the Woman's Soul (Ch...   
15  1567407781       The Witchfinder (Amos Walker Mystery Series)   
16  1575663937  More Cunning Than Man: A Social History of Rat...   
17  1881320189                      Goodbye to the Buttermilk Sky   
18  0440234743                                      The Testament   
19  0452264464               Beloved (Plume Contemporary Fiction)   
20  0609804618  Our Dumb Century: The Onion Presents 100 Years...   

                  author  year                    publisher  
0     Mark P. O. Morford  2002      Oxford University Press  
1   Richard Bruce Wright  2001        HarperFlamingo Canada  
2           Carlo D'Este  1991              HarperPerennial  
3       Gina Bari Kolata  1999         Farrar Straus Giroux  
4        E. J. W. Barber  1999   W. W. Norton &amp; Company  
5                Amy Tan  1991             Putnam Pub Group  
6          Robert Cowley  2000     Berkley Publishing Group  
7            Scott Turow  1993                   Audioworks  
8        David Cordingly  1996                 Random House  
9            Ann Beattie  2002                     Scribner  
10  David Adams Richards  1988              Emblem Editions  
11            Adam Lebor  2000                Citadel Press  
12           Sheila Heti  2004        House of Anansi Press  
13          R. J. Kaiser  1999                   Mira Books  
14         Jack Canfield  1998        Health Communications  
15     Loren D. Estleman  1998     Brilliance Audio - Trade  
16    Robert Hendrickson  1999  Kensington Publishing Corp.  
17          Julia Oliver  1994               River City Pub  
18          John Grisham  1999                         Dell  
19         Toni Morrison  1994                        Plume  
20             The Onion  1999           Three Rivers Press

In [33]:
users.loc[0:20]

user_id                                location   age
0         1                      nyc, new york, usa   NaN
1         2               stockton, california, usa  18.0
2         3         moscow, yukon territory, russia   NaN
3         4               porto, v.n.gaia, portugal  17.0
4         5      farnborough, hants, united kingdom   NaN
5         6           santa monica, california, usa  61.0
6         7                     washington, dc, usa   NaN
7         8                timmins, ontario, canada   NaN
8         9              germantown, tennessee, usa   NaN
9        10              albacete, wisconsin, spain  26.0
10       11          melbourne, victoria, australia  14.0
11       12             fort bragg, california, usa   NaN
12       13             barcelona, barcelona, spain  26.0
13       14                   mediapolis, iowa, usa   NaN
14       15                calgary, alberta, canada   NaN
15       16            albuquerque, new mexico, usa   NaN
16       17               chesapeake, virginia, usa   NaN
17       18  rio de janeiro, rio de janeiro, brazil  25.0
18       19                               weston, ,  14.0
19       20            langhorne, pennsylvania, usa  19.0
20       21          ferrol / spain, alabama, spain  46.0

In [36]:
ratings.loc[0:20]

user_id        ISBN  rating
0    276725  034545104X       0
1    276726  0155061224       5
2    276727  0446520802       0
3    276729  052165615X       3
4    276729  0521795028       6
5    276733  2080674722       0
6    276736  3257224281       8
7    276737  0600570967       6
8    276744  038550120X       7
9    276745   342310538      10
10   276746  0425115801       0
11   276746  0449006522       0
12   276746  0553561618       0
13   276746  055356451X       0
14   276746  0786013990       0
15   276746  0786014512       0
16   276747  0060517794       9
17   276747  0451192001       0
18   276747  0609801279       0
19   276747  0671537458       9
20   276747  0679776818       8

In [38]:
#number of user that given ratings
ratings['user_id'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: user_id, Length: 105283, dtype: int64

In [39]:
# extracting the number of user who have given more than 200 ratings
x = ratings['user_id'].value_counts() > 200
y = x[x].index  #user_ids
print(y.shape)
ratings = ratings[ratings['user_id'].isin(y)]

(899,)


In [42]:
type(x)  # extracting data of one column gives us panda series

pandas.core.series.Series

In [41]:
type(y)

pandas.core.indexes.numeric.Int64Index

<p>So 899 users are there who have given 5.2 lakh rating and this we want. Now we will merge ratings with books on basis of ISBN so that we will get the rating of each user on each book id and the user who has not rated that book id the value will be zero.</p>

In [50]:
rating_with_books = ratings.merge(books, on='ISBN')
rating_with_books.head(50)

user_id        ISBN  rating  \
0    277427  002542730X      10   
1      3363  002542730X       0   
2     11676  002542730X       6   
3     12538  002542730X      10   
4     13552  002542730X       0   
5     16795  002542730X       0   
6     24194  002542730X       0   
7     25981  002542730X       0   
8     26535  002542730X       0   
9     28204  002542730X       0   
10    30533  002542730X       0   
11    31826  002542730X       0   
12    41084  002542730X       7   
13    43246  002542730X       0   
14    44728  002542730X       0   
15    51883  002542730X       0   
16    52584  002542730X      10   
17    55548  002542730X       0   
18    60244  002542730X       0   
19    71712  002542730X       6   
20    77809  002542730X       0   
21    80538  002542730X       7   
22    81045  002542730X       0   
23    82407  002542730X       0   
24    82831  002542730X       0   
25    86243  002542730X       7   
26    88733  002542730X       0   
27   110934  002542730X       6   
28   111947  002542730X       0   
29   112026  002542730X       0   
30   113270  002542730X       5   
31   119575  002542730X       0   
32   126604  002542730X       0   
33   128696  002542730X       0   
34   128835  002542730X       8   
35   130474  002542730X       0   
36   131837  002542730X       0   
37   150979  002542730X       8   
38   155147  002542730X       0   
39   159858  002542730X       0   
40   164096  002542730X       8   
41   165308  002542730X       0   
42   170518  002542730X       0   
43   172030  002542730X       0   
44   173291  002542730X       7   
45   173415  002542730X       0   
46   174216  002542730X       5   
47   174304  002542730X       0   
48   178667  002542730X       0   
49   179733  002542730X       0   

                                                title             author  \
0   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
5   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
6   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
7   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
8   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
9   Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
10  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
11  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
12  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
13  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
14  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
15  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
16  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
17  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
18  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
19  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
20  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
21  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
22  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
23  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
24  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
25  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
26  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
27  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
28  Politicall

In [51]:
rating_with_books.shape

(487671, 7)

<p>Now dataframe size has decreased and we have 4.8 lakh because when we merge the dataframe, all the book id-data we were not having. Now we will count the rating of each book so we will group data based on title and aggregate based on rating.</p>

In [52]:
number_rating = rating_with_books.groupby('title')['rating'].count().reset_index()
number_rating.rename(columns= {'rating':'number_of_ratings'}, inplace=True)
final_rating = rating_with_books.merge(number_rating, on='title')
final_rating.shape
final_rating = final_rating[final_rating['number_of_ratings'] >= 50]
final_rating.drop_duplicates(['user_id','title'], inplace=True)

we will create a pivot table where columns will be user ids, the index will be book title and the value is ratings. And the user id who has not rated any book will have value as NAN so impute it with zero

In [53]:
book_pivot = final_rating.pivot_table(columns='user_id', index='title', values="rating")
book_pivot.fillna(0, inplace=True)

In [56]:
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

# Modelling 

<p>We have prepared our dataset for modeling. we will use the nearest neighbors algorithm which is the same as K nearest which is used for clustering based on euclidian distance.</p>

<p>But here in the pivot table, we have lots of zero values and on clustering, this computing power will increase to calculate the distance of zero values so we will convert the pivot table to the sparse matrix and then feed it to the model.</p>

In [58]:
from scipy.sparse import csr_matrix
book_sparse = csr_matrix(book_pivot)

In [63]:
print(book_sparse)

  (0, 0)	9.0
  (0, 16)	8.0
  (0, 37)	9.0
  (0, 42)	8.0
  (0, 59)	7.0
  (0, 135)	9.0
  (0, 156)	10.0
  (0, 170)	9.0
  (0, 187)	8.0
  (0, 220)	10.0
  (0, 221)	10.0
  (0, 245)	10.0
  (0, 277)	10.0
  (0, 359)	10.0
  (0, 370)	9.0
  (0, 389)	10.0
  (0, 477)	9.0
  (0, 491)	7.0
  (0, 507)	9.0
  (0, 554)	10.0
  (0, 604)	9.0
  (0, 611)	8.0
  (0, 695)	9.0
  (0, 734)	9.0
  (0, 762)	8.0
  :	:
  (740, 484)	8.0
  (740, 681)	9.0
  (741, 16)	8.0
  (741, 26)	7.0
  (741, 86)	10.0
  (741, 111)	5.0
  (741, 150)	6.0
  (741, 233)	7.0
  (741, 290)	8.0
  (741, 303)	9.0
  (741, 322)	10.0
  (741, 367)	10.0
  (741, 384)	9.0
  (741, 446)	7.0
  (741, 450)	9.0
  (741, 485)	7.0
  (741, 498)	10.0
  (741, 505)	8.0
  (741, 544)	10.0
  (741, 592)	8.0
  (741, 700)	10.0
  (741, 713)	8.0
  (741, 745)	10.0
  (741, 830)	9.0
  (741, 880)	8.0


Now we will train the nearest neighbors algorithm. here we need to specify an algorithm which is brute means find the distance of every point to every other point.

In [64]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm='brute')
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

see whether it is suggesting books or not. we will find the nearest neighbors to the input book id and after that, we will print the top 5 books which are closer to those books. It will provide us distance and book id at that distance. let us pass harry potter which is at 237 indexes.

In [69]:
distances, suggestions = model.kneighbors(book_pivot.iloc[1, :].values.reshape(1, -1))

In [70]:
for i in range(len(suggestions)):
  print(book_pivot.index[suggestions[i]])

Index(['1st to Die: A Novel', 'Exclusive', 'The Cradle Will Fall',
       'The Clinic (Alex Delaware Novels (Paperback))', 'No Safe Place'],
      dtype='object', name='title')
